cmd/jailrun/server.go

```diff
@@ -0,0 +1,112 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"os"
+	"runtime"
+	"sync"
+
+	"golang.org/x/sys/unix"
+	"github.com/redpwn/jail/pow"
+	"inet.af/netaddr"
+)
+
+var connMu sync.Mutex
+var connCountPerIp = make(map[netaddr.IP]uint32)
+var	connCount uint32
+
+func connInc(cfg *jailConfig, ip netaddr.IP) bool {
+	connMu.Lock()
+	defer connMu.Unlock()
+	if (cfg.Conns > 0 && connCount >= cfg.Conns) || (cfg.ConnsPerIp > 0 && connCountPerIp[ip] >= cfg.ConnsPerIp) {
+		return false
+	}
+	connCountPerIp[ip] += 1
+	connCount += 1
+	return true
+}
+
+func connDec(ip netaddr.IP) {
+	connMu.Lock()
+	defer connMu.Unlock()
+	connCountPerIp[ip] -= 1
+	connCount -= 1
+}
+
+func runConn(cfg *jailConfig, c net.Conn, errCh chan error) {
+	defer c.Close()
+	addr := c.RemoteAddr().(*net.TCPAddr)
+	log.Printf("connection: %s", addr)
+	ip, ok := netaddr.FromStdIP(addr.IP)
+	if !ok {
+		return
+	}
+	if !connInc(cfg, ip) {
+		log.Printf("connection: %s: limit reached", addr)
+		return
+	}
+	defer connDec(ip)
+	chal := pow.GenerateChallenge(cfg.Pow)
+	r := bufio.NewReader(io.LimitReader(c, 1024))
+	c.Write([]byte(fmt.Sprintf("proof of work: curl -sSfL https://pwn.red/pow | sh -s %s\nsolution: ", chal)))
+	s, err := r.ReadString('\n')
+	if err != nil {
+		return
+	}
+	correct, err := chal.Check(s)
+	if err != nil || !correct {
+		log.Printf("connection: %s: bad pow", addr)
+		c.Write([]byte("incorrect proof of work\n"))
+		return
+	}
+	log.Printf("connection: %s: forwarding", addr)
+	d, err := net.Dial("tcp", fmt.Sprintf(":%d", cfg.Port+1))
+	if err != nil {
+		errCh <- err
+		return
+	}
+	defer d.Close()
+	eofCh := make(chan struct{})
+	go func() {
+		io.Copy(c, d)
+		eofCh <- struct{}{}
+	}()
+	go func() {
+		io.Copy(d, c)
+		eofCh <- struct{}{}
+	}()
+	<-eofCh
+}
+
+func startServer(cfg *jailConfig, errCh chan error) {
+	l, err := net.Listen("tcp", fmt.Sprintf(":%d", cfg.Port))
+	if err != nil {
+		errCh <- err
+		return
+	}
+	log.Printf("listening on %d", cfg.Port)
+	defer l.Close()
+	for {
+		c, err := l.Accept()
```

Comment:
We should try to not even accept connections when the global limit is exceeded, instead of accepting and then immediately closing, so connections can get buffered or otherwise handled by the network and host networking stack. Maybe also move all connection limiting logic onto the `Accept()` thread?